process/sync/shardForkDetector.go

```diff
@@ -69,51 +78,148 @@ func (sfd *shardForkDetector) AddHeader(
 		return err
 	}
 
-	sfd.activateForcedForkIfNeeded(header, state)
+	if header.GetNonce() > sfd.highestNonceReceived() {
+		log.Debug("forkDetector.AddHeader.setHighestNonceReceived",
+			"nonce", header.GetNonce(),
+			"highest nonce received", sfd.highestNonceReceived())
+		sfd.setHighestNonceReceived(header.GetNonce())
+	}
 
-	isHeaderReceivedTooLate := sfd.isHeaderReceivedTooLate(header, state, process.ShardBlockFinality)
-	if isHeaderReceivedTooLate {
-		state = process.BHReceivedTooLate
+	if state == process.BHProposed {
+		sfd.activateForcedForkOnConsensusStuckIfNeeded(header, state)
```

Comment:
code looks really close to metaForkDetector.

Reply:
Yep, good hint. Extracted in baseForkDetector